### PR TITLE
Adds parameter lookup function for kinematics plugins

### PR DESCRIPTION
**Pull request by henhenhen**
Some kinematics plugins access parameters that are defined in the private namespace of their node.
With multiple nodes, each node needs to be initialized with the same parameters which can be
very inconvenient in some setups.
This commit resolves the issue by enabling plugins to query parameters from the
namespace 'robot_description_kinematics' so that they can be defined in the kinematics.yaml either under the corresponding group or directly.

This PR helps to resolve issues [with multiple arms](https://github.com/ros-industrial/universal_robot/issues/247) and when launching the plugin with [rviz and the moveit commander simultaneously](https://github.com/ros-industrial/universal_robot/issues/256) (also see [here](https://github.com/ros-planning/moveit/issues/545) and [here](https://bitbucket.org/traclabs/trac_ik/issues/25/unable-to-set-solver_type-in-launch-file)), although the plugins will actually have to make use of the new method.

Example:
In our UR5 setup we add a prefix 'ur5_' to each joint name in the robot description.
When using the UR5KinematicsPlugin from ur_kinematics we have to define the parameter 'arm_prefix' at every launch within the private namespace of each node using the kinematics plugin.
With this PR and [this minor change](https://github.com/TAMS-Group/universal_robot/commit/e384a27cd6ae149826f5bb474f462d15d3f9330e) in ur_kinematics we just need to set the parameter once inside the kinematics.yaml under the group arm.

@v4hn

### CODE REVIEW

**Comment by v4hn** [comment]
This is the best solution @henhenhen and I came up with to finally resolve all the linked issues/bugs.
We will have to change the parameter lookup in all common IK plugins, but I guess it's worth the trouble.
We demonstrated it's not much work in our patch for `ur_kinematics`.

I just ran abi-compliance-checker and it agrees with me in that the change is API/ABI stable.
Thus I would like to port it back to indigo too.

Feedback/Approval @davetcoleman @bmagyar @jrgnicho ?

File: moveit_core/kinematics_base/include/moveit/kinematics_base/kinematics_base.h [hunk 1]
```diff
@@ -598,6 +600,36 @@ class KinematicsBase
   std::map<int, double> redundant_joint_discretization_;
   std::vector<DiscretizationMethod> supported_methods_;
 
+  /**
+   * @brief Enables kinematics plugins access to parameters that are defined
+   * for the 'robot_description_kinematics' namespace.
```
Comment (davetcoleman) [inline moveit_core/kinematics_base/include/moveit/kinematics_base/kinematics_base.h]: Could you add a comment explaining how it tried to find the param in the root namespace first, then in the group name subgroup second?

File: moveit_core/kinematics_base/include/moveit/kinematics_base/kinematics_base.h [hunk 2]
```diff
@@ -598,6 +600,36 @@ class KinematicsBase
...
+    ros::NodeHandle nh;
+    if (nh.hasParam("robot_description_kinematics/" + param))
+    {
+      val = nh.param("robot_description_kinematics/" + param, default_val);
+      return true;
+    }
+
+    if (nh.hasParam("robot_description_kinematics/" + group_name_ + "/" + param))
+    {
+      val = nh.param("robot_description_kinematics/" + group_name_ + "/" + param, default_val);
```
Comment (davetcoleman) [inline moveit_core/kinematics_base/include/moveit/kinematics_base/kinematics_base.h]: are we sure we want the ordering of looking for the parameter like this, with group second?

Comment (v4hn) [reply moveit_core/kinematics_base/include/moveit/kinematics_base/kinematics_base.h]: No I'm not sure. I'm actually surprised to see it this way around now that you mention it...
In my opinion more specific places should go first and the order should be
- `~`
- `robot_description_kinematics/group_name/`
- `robot_description_kinematics`

@henhenhen could you please update the request (or provide insights that justify the current order)?

File: moveit_core/kinematics_base/include/moveit/kinematics_base/kinematics_base.h (hunk 1 above)
Comment (henhenhen) [reply moveit_core/kinematics_base/include/moveit/kinematics_base/kinematics_base.h]: https://github.com/ros-planning/moveit/pull/701/commits/45a35c83951d0bab04db7318554791461d841f1c adds some explanation for the lookup order.

File: moveit_core/kinematics_base/include/moveit/kinematics_base/kinematics_base.h (hunk 2 above)
Comment (henhenhen) [reply moveit_core/kinematics_base/include/moveit/kinematics_base/kinematics_base.h]: Actually the order should have been like @v4hn mentioned. 
Private namespace as the most specific and prior default should be queried first.
Also subgroup should be checked before the group to allow group-wise specifications.

I changed it with commit https://github.com/ros-planning/moveit/pull/701/commits/53cb47064a49c13db9a6ba1524a27f91f63fb01b

**Pull request merged**